markovflow/sde/sde_utils.py

```diff
@@ -0,0 +1,154 @@
+
+    :param sde: SDE to be linearized.
+    :param q_mean: mean of Gaussian over states with shape (n_batch, num_states, state_dim).
+    :param q_covar: covariance of Gaussian over states with shape (n_batch, num_states, state_dim, state_dim).
+
+    :return: the expectation value with shape (n_batch, num_states, state_dim).
+    """
+    fx = lambda x: sde.drift(x=x, t=tf.zeros(x.shape[0], 1))
+
+    n_batch, n_states, state_dim = q_mean.shape
+    q_mean = tf.reshape(q_mean, (-1, state_dim))
+    q_covar = tf.reshape(q_covar, (-1, state_dim, state_dim))
+
+    val = mvnquad(fx, q_mean, q_covar, H=10)
+
+    val = tf.reshape(val, (n_batch, n_states, state_dim))
+    return val
+
+
+def E_sde_drift_gradient(sde: SDE, q_mean: TensorType, q_covar: TensorType) -> TensorType:
+    """
+     Calculates the Expectation of the gradient of the SDE's drift under the provided Gaussian over states
+
+    ..math:: E_q(.)[f'(x(t))]
+
+    :param sde: SDE to be linearized.
+    :param q_mean: mean of Gaussian over states with shape (n_batch, num_states, state_dim).
+    :param q_covar: covariance of Gaussian over states with shape (n_batch, num_states, state_dim, state_dim).
+
+    :return: the expectation value with shape (n_batch, num_states, state_dim).
+    """
+    n_batch, n_states, state_dim = q_mean.shape
+    q_mean = tf.reshape(q_mean, (-1, state_dim))
+    q_covar = tf.reshape(q_covar, (-1, state_dim, state_dim))
+    val = mvnquad(sde.sde_drift_gradient, q_mean, q_covar, H=10)
+
+    val = tf.reshape(val, (n_batch, n_states, state_dim))
+    return val
+
+
+def linearize_sde(sde: SDE, q_mean: TensorType, q_covar: TensorType, initial_mean: TensorType,
+                  initial_chol_covariance: TensorType, process_chol_covariances: TensorType) -> StateSpaceModel:
+    """
+    Linearizes the SDE (with fixed diffusion) on the basis of the Gaussian over states
+
+    ..math:: q(\cdot) \sim N(q_{mean}, q_{covar})
+
+    ..math:: A_{i}^{*} = E_{q(.)}[d f(x)/ dx]
+    ..math:: b_{i}^{*} = E_{q(.)}[f(x)] - A_{i}^{*}  E_{q(.)}[x]
+
+    :param sde: SDE to be linearized.
+    :param q_mean: mean of Gaussian over states with shape (n_batch, num_states, state_dim).
+    :param q_covar: covariance of Gaussian over states with shape (n_batch, num_states, state_dim, state_dim).
+    :param linearize_points: points used for the piecewise linearization ``[num_states,]``.
+    :param initial_mean: The initial mean, with shape ``[n_batch, state_dim]``.
+    :param initial_chol_covariance: Cholesky of the initial covariance, with shape ``[n_batch, state_dim, state_dim]``.
+    :param process_chol_covariances: Cholesky of the noise covariance matrices, with shape
+        ``[n_batch, num_states, state_dim, state_dim]``.
+
+    :return: the state-space model of the linearized SDE.
+    """
+
+    E_f = E_sde_drift(sde, q_mean, q_covar)
+    E_x = q_mean
+
+    A = E_sde_drift_gradient(sde, q_mean, q_covar)
+    b = E_f - A * E_x
+    A = tf.linalg.diag(A)
```

Comment:
This is only for 1d, which is fine for now, but then mention that in the docstrings + assert